ddtrace/tracer/writer.go

```diff
@@ -201,22 +208,36 @@ func (h *logTraceWriter) encodeSpan(s *span) {
 			continue
 		}
 		if first {
-			h.buf.WriteByte('"')
 			first = false
 		} else {
-			h.buf.WriteString(`,"`)
+			h.buf.WriteString(`,`)
+		}
+		m, err := json.Marshal(k)
+		if err != nil {
+			log.Error("Error encoding metrics key %q: %v", k, err)
 		}
-		h.buf.WriteString(k)
-		h.buf.WriteString(`":`)
+		h.buf.Write(m)
+		h.buf.WriteString(`:`)
 		h.buf.Write(encodeFloat(scratch[:0], v))
 	}
 	h.buf.WriteString(`},"start":`)
 	h.buf.Write(strconv.AppendInt(scratch[:0], s.Start, 10))
 	h.buf.WriteString(`,"duration":`)
 	h.buf.Write(strconv.AppendInt(scratch[:0], s.Duration, 10))
-	h.buf.WriteString(`,"service":"`)
-	h.buf.WriteString(s.Service)
-	h.buf.WriteString(`"}`)
+	h.buf.WriteString(`,"service":`)
+	h.marshalString(s.Service)
+	h.buf.WriteString(`}`)
+}
+
+// marshalString will sanitize strings that may hold characters
+// that would render invalid JSON.
```

Comment:
This text explains the current PR and the motivations. Here, we must describe what this function does:
```go
// marshalString marshals the string str as JSON into the writer's buffer.
```
That's all :) The bug details are irrelevant here because we go by the presumption that things work correctly in documentation.

Reply:
OTOH, if it read:

```golang
// marshalString marshals the string str as JSON into the writer's buffer. Should be used whenever writing non-constant string data to ensure correct sanitization.
```

You've described what it does, when it's needed (which is otherwise unclear), and why it's important. The what will be fairly clear from the implementation anyway.

That said, I don't maintain this repo, so I don't have skin in the game 😁

Reply:
Good point.

> That said, I don't maintain this repo, so I don't have skin in the game

Doesn't matter. All feedback is welcome!